File: answers/leetcode/Binary Tree Level Order Traversal II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """

        level_list = []
        self.dfs(0, root, level_list)
        return level_list[: : -1]

    def dfs(self, level, root, level_list):
        if None == root:
            return
        while len(level_list) <= level:
            level_list.append([])
        level_list[level].append(root.val)
        self.dfs(level + 1, root.left, level_list)
        self.dfs(level + 1, root.right, level_list)

File: answers/other/answers/leetcode/Binary Tree Level Order Traversal II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	vector<vector<int> > levelOrderBottom(TreeNode *root) {
		levelOrderBottom(root, 0);
		std::reverse(level_order.begin(), level_order.end());
		return level_order;
	}
	void levelOrderBottom(TreeNode *root, int level) {
		if (NULL == root) {
			return;
		}
		if (level_order.size() <= level) {
			level_order.resize(level + 1);
		}
		level_order[level].push_back(root->val);
		levelOrderBottom(root->left, level + 1);
		levelOrderBottom(root->right, level + 1);
	}
	vector<vector<int> > level_order;
};
